server/session/world.go

```diff
@@ -22,90 +22,137 @@ import (
+		for pos, b := range ch.BlockEntities() {
+			if n, ok := b.(world.NBTer); ok && ch.SubIndex(int16(pos.Y())) == ind {
+				d := n.EncodeNBT()
+				d["x"], d["y"], d["z"] = int32(pos[0]), int32(pos[1]), int32(pos[2])
+				_ = enc.Encode(d)
+			}
+		}
+
+		entry := protocol.SubChunkEntry{
+			Result:        protocol.SubChunkResultSuccess,
+			RawPayload:    append(serialisedSubChunk, blockEntityBuf.Bytes()...),
+			HeightMapType: subMapType,
+			HeightMapData: subMap,
+			Offset:        offset,
+		}
+		if s.conn.ClientCacheEnabled() {
+			hash := xxhash.Sum64(serialisedSubChunk)
+			if !s.openTransaction(hash, serialisedSubChunk) {
+				// Failed to open the transaction, so just stop here.
+				return
+			}
+			entry.BlobHash = hash
+			entry.RawPayload = blockEntityBuf.Bytes()
+		}
+		entries = append(entries, entry)
+	}
+	s.writePacket(&packet.SubChunk{
+		Dimension:       int32(w.Dimension().EncodeDimension()),
+		Position:        protocol.SubChunkPos(center),
+		CacheEnabled:    s.conn.ClientCacheEnabled(),
+		SubChunkEntries: entries,
 	})
 }
 
-// sendNetworkChunk sends a network encoded chunk to the client.
-func (s *Session) sendNetworkChunk(pos world.ChunkPos, c *chunk.Chunk, blockEntities map[cube.Pos]world.Block) {
-	data := chunk.Encode(c, chunk.NetworkEncoding)
+// ViewSkeletonChunk ...
+func (s *Session) ViewSkeletonChunk(pos world.ChunkPos, c *chunk.Chunk) {
+	biomes := chunk.EncodeBiomes(c, chunk.NetworkEncoding)
+	if !s.conn.ClientCacheEnabled() {
+		s.writePacket(&packet.LevelChunk{
+			SubChunkRequestMode: protocol.SubChunkRequestModeLimited,
+			Position:            protocol.ChunkPos(pos),
+			HighestSubChunk:     uint16(len(c.Sub())), // This is always going to be the highest sub-chunk, anyway.
+			RawPayload:          biomes,
+		})
+		return
+	}
 
-	for i := range data.SubChunks {
-		_, _ = s.chunkBuf.Write(data.SubChunks[i])
+	if hash := xxhash.Sum64(biomes); s.openTransaction(hash, biomes) {
+		s.writePacket(&packet.LevelChunk{
+			SubChunkRequestMode: protocol.SubChunkRequestModeLimited,
+			Position:            protocol.ChunkPos(pos),
+			HighestSubChunk:     uint16(len(c.Sub())), // This is always going to be the highest sub-chunk, anyway.
+			BlobHashes:          []uint64{hash},
+			CacheEnabled:        true,
+		})
 	}
-	_, _ = s.chunkBuf.Write(data.Biomes)
+}
 
-	// Length of 1 byte for the border block count.
-	s.chunkBuf.WriteByte(0)
+// openTransaction opens a chunk transaction for a given blob.
+func (s *Session) openTransaction(hash uint64, blob []byte) bool {
+	s.blobMu.Lock()
+	defer s.blobMu.Unlock()
 
-	enc := nbt.NewEncoderWithEncoding(s.chunkBuf, nbt.NetworkLittleEndian)
-	for bp, b := range blockEntities {
-		if n, ok := b.(world.NBTer); ok {
-			d := n.EncodeNBT()
-			d["x"], d["y"], d["z"] = int32(bp[0]), int32(bp[1]), int32(bp[2])
-			_ = enc.Encode(d)
-		}
+	s.openChunkTransactions = append(s.openChunkTransactions, map[uint64]struct{}{hash: {}})
```

Comment:
I don't think this will work as intended. This system expects a map[uint64]struct{} with all hashes of every sub-chunk in a chunk. Every chunk load tick, `8 - len(s.openChunkTransactions)` chunks are loaded, which means that with this change a single chunk could prevent any other chunks from being sent because the chunk has more than 8 subchunks.

Reply:
Hmm, what do you suggest I do instead?

Reply:
I would open a single transaction containing a map[uint64]struct{} with the hashes that came from all the `Offset` requested together.

Reply:
Essentially this needs to be an indication of how many chunks are currently in the process of being loaded client-side. If this number goes up too high, we need to slow down chunk sending.